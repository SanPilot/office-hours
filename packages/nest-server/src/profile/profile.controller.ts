import { Controller, Get, Res, Query, UseGuards } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { Response } from 'express';
import { Connection } from 'typeorm';
import { User as UserModel } from './user.entity';
import { pick } from 'lodash';
import { GetProfileResponse } from '@template/common';
import { JwtAuthGuard } from './jwt-auth.guard';
import { User } from './user.decorator';

@Controller('profile')
export class ProfileController {
  constructor(private connection: Connection, private jwtService: JwtService) {}

  @UseGuards(JwtAuthGuard)
  @Get()
  async get(
    @User(['courses', 'courses.course']) user: UserModel,
  ): Promise<GetProfileResponse> {
    const courses = user.courses.map((userCourse) => {
      return {
        course: {
          id: userCourse.courseId,
          name: userCourse.course.name,
        },
        role: userCourse.role,
      };
    });

    const userResponse = pick(user, ['id', 'email', 'name', 'photoURL']);
    return { ...userResponse, courses };
  }

  // TODO handle the khoury flow for real.
  @Get('/entry')
  enterFromKhoury(@Res() res: Response, @Query('userId') userId: number): void {
    const token = this.jwtService.sign({ userId });
    res.cookie('auth_token', token).status(200).send();
  }
}
