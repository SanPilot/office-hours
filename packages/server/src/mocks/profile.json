{
  "id": 258,
  "email": "stenzel.w@husky.neu.edu",
  "name": "Will Stenzel",
  "photo_url": "https://prod-web.neu.edu/wasapp/EnterprisePhotoService/PhotoServlet?vid=CCS&er=da30ed4153a22ec0661fe08d52bf29ae29220b5a507f641dcc0a6af3862935bb43b5e2dbadff91cd1e95bd73fb867c75a7a823a4b640ba91",
  "courses": [
    {
      "id": 169,
      "name": "CS 2500",
      "role": "ta"
    },
    {
      "id": 21,
      "name": "CS 4500",
      "role": "student"
    }
  ]
}
