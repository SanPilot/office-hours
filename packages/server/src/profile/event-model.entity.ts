import { EventType } from '@koh/common';
import {
  BaseEntity,
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { UserCourseModel } from './user-course.entity';

@Entity('user_course_model')
export class EventModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  time: Date;

  @Column({ type: 'enum', enum: EventType })
  eventType: EventType;

  @ManyToOne((type) => UserCourseModel, (ucm) => ucm.events)
  @JoinColumn({ name: 'userCourseId' })
  user: UserCourseModel;

  @Column({ nullable: true })
  userCourseId: number;
}
