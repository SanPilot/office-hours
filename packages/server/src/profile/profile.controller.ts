import {
  Controller,
  Get,
  Res,
  Query,
  UseGuards,
  Patch,
  Body,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { Response } from 'express';
import { Connection } from 'typeorm';
import { UserModel } from './user.entity';
import { pick } from 'lodash';
import { GetProfileResponse, UpdateProfileParams } from '@template/common';
import { JwtAuthGuard } from './jwt-auth.guard';
import { User } from './user.decorator';
import { NotificationService } from '../notification/notification.service';

@Controller('profile')
export class ProfileController {
  constructor(
    private connection: Connection,
    private jwtService: JwtService,
    private notifService: NotificationService,
  ) {}

  @UseGuards(JwtAuthGuard)
  @Get()
  async get(
    @User(['courses', 'courses.course', 'phoneNotif']) user: UserModel,
  ): Promise<GetProfileResponse> {
    const courses = user.courses.map((userCourse) => {
      return {
        course: {
          id: userCourse.courseId,
          name: userCourse.course.name,
        },
        role: userCourse.role,
      };
    });

    const userResponse = pick(user, [
      'id',
      'email',
      'name',
      'photoURL',
      'desktopNotifsEnabled',
      'phoneNotifsEnabled',
    ]);
    return {
      ...userResponse,
      courses,
      phoneNumber: user.phoneNotif?.phoneNumber,
    };
  }

  @UseGuards(JwtAuthGuard)
  @Patch()
  async patch(
    @Body() userPatch: UpdateProfileParams,
    @User(['courses', 'courses.course', 'phoneNotif']) user: UserModel,
  ): Promise<GetProfileResponse> {
    user = Object.assign(user, userPatch);
    if (user.phoneNotifsEnabled && userPatch.phoneNumber) {
      await this.notifService.registerPhone(userPatch.phoneNumber, user.id);
    }
    await user.save();

    const courses = user.courses.map((userCourse) => {
      return {
        course: {
          id: userCourse.courseId,
          name: userCourse.course.name,
        },
        role: userCourse.role,
      };
    });

    const userResponse = pick(user, [
      'id',
      'email',
      'name',
      'photoURL',
      'desktopNotifsEnabled',
      'phoneNotifsEnabled',
    ]);
    return {
      ...userResponse,
      courses,
      phoneNumber: user.phoneNotif?.phoneNumber,
    };
  }

  // TODO handle the khoury flow for real.
  @Get('/entry')
  enterFromKhoury(@Res() res: Response, @Query('userId') userId: number): void {
    const token = this.jwtService.sign({ userId });
    res.cookie('auth_token', token).redirect(302, '/');
  }
}
