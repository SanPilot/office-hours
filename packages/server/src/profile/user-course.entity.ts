import { Role } from '@koh/common';
import { Exclude } from 'class-transformer';
import {
  BaseEntity,
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { CourseModel } from '../course/course.entity';
import { EventModel } from './event-model.entity';
import { UserModel } from './user.entity';

@Entity('user_course_model')
export class UserCourseModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => UserModel, (user) => user.courses)
  @JoinColumn({ name: 'userId' })
  user: UserModel;

  @Column({ nullable: true })
  userId: number;

  @ManyToOne((type) => CourseModel, (course) => course.userCourses)
  @JoinColumn({ name: 'courseId' })
  course: CourseModel;

  @Column({ nullable: true })
  courseId: number;

  @Column({ type: 'enum', enum: Role, default: Role.STUDENT })
  role: Role;

  @Exclude()
  @OneToMany((type) => EventModel, (event) => event.user)
  events: EventModel[];
}
