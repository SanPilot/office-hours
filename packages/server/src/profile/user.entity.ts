import { Exclude } from 'class-transformer';
import { INSIGHTS_MAP } from '../insights/insight-objects';
import {
  AfterLoad,
  BaseEntity,
  Column,
  Entity,
  ManyToMany,
  OneToMany,
  OneToOne,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { DesktopNotifModel } from '../notification/desktop-notif.entity';
import { PhoneNotifModel } from '../notification/phone-notif.entity';
import { QueueModel } from '../queue/queue.entity';
import { EventModel } from './event-model.entity';
import { UserCourseModel } from './user-course.entity';

@Entity('user_model')
export class UserModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column('text')
  email: string;

  @Column('text', { nullable: true })
  firstName: string;

  @Column('text', { nullable: true })
  lastName: string;

  @Column('text', { nullable: true })
  photoURL: string;

  @OneToMany((type) => UserCourseModel, (ucm) => ucm.user)
  @Exclude()
  courses: UserCourseModel[];

  @Column({ type: 'boolean', default: false })
  @Exclude()
  desktopNotifsEnabled: boolean; // Does user want notifications sent to their desktops?

  @Column({ type: 'boolean', default: false })
  @Exclude()
  phoneNotifsEnabled: boolean; // Does user want notifications sent to their phone?

  @OneToMany((type) => DesktopNotifModel, (notif) => notif.user)
  @Exclude()
  desktopNotifs: DesktopNotifModel[];

  @OneToOne((type) => PhoneNotifModel, (notif) => notif.user)
  @Exclude()
  phoneNotif: PhoneNotifModel;

  @Exclude()
  @ManyToMany((type) => QueueModel, (queue) => queue.staffList)
  queues: QueueModel[];

  @Exclude()
  @OneToMany((type) => EventModel, (event) => event.user)
  events: EventModel[];

  @Exclude()
  @Column({ type: 'simple-array', nullable: true, default: [] })
  hideInsights: string[];

  insights: string[];

  @AfterLoad()
  computeInsights() {
    if (this.hideInsights) {
      const insightNames = Object.keys(INSIGHTS_MAP);
      this.insights = insightNames.filter(
        (name) => !this.hideInsights.includes(name),
      );
    }
  }

  name: string;

  @AfterLoad()
  setFullNames() {
    this.name = this.firstName + ' ' + this.lastName;
  }
}
