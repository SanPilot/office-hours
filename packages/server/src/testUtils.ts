import { createConnection, Connection } from "../../ormconfig";
import { init } from "./server";
import path from "path";
import { Server } from "@hapi/hapi";

export function setupServerTest(): () => Server {
  let server: Server;

  beforeEach(async () => {
    server = await init();
  });

  afterEach(async () => {
    await server.stop();
  });

  return () => server;
}

export function setupDBTest() {
  let connection: Connection;
  beforeAll(async () => {
    connection = await createConnection({
      type: "postgres",
      url: "postgres://postgres@localhost:5432/test",
      synchronize: true,
      logging: false,
      entities: [path.resolve(__dirname, "entity/**/*{.js,.ts}")],
      migrations: [path.resolve(__dirname, "migration/**/*{.js,.ts}")],
      subscribers: [path.resolve(__dirname, "subscriber/**/*{.js,.ts}")],
    });
  });

  beforeEach(async () => {
    await connection.synchronize(true);
  });

  afterAll(async () => {
    await connection.close();
  });
}
