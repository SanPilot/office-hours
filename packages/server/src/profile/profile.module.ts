import { Module } from '@nestjs/common';
import { ProfileController } from './profile.controller';
import { NotificationModule } from '../notification/notification.module';

@Module({
  imports: [NotificationModule],
  controllers: [ProfileController],
})
export class ProfileModule {}
