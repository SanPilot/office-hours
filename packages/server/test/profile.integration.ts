import {
  StudentCourseFactory,
  UserFactory,
  CourseFactory,
} from './util/factories';
import { setupIntegrationTest } from './util/testUtils';
import { ProfileModule } from '../src/profile/profile.module';

describe('Profile Integration', () => {
  const supertest = setupIntegrationTest(ProfileModule);

  describe('GET /profile', () => {
    it('returns the logged-in user profile', async () => {
      const user = await UserFactory.create();
      const fundies = await CourseFactory.create({ name: 'CS 2500' });
      await StudentCourseFactory.create({ course: fundies, user });

      const res = await supertest({ userId: user.id })
        .get('/profile')
        .expect(200);
      expect(res.body).toMatchSnapshot();
    });

    it('returns 401 when not logged in', async () => {
      await UserFactory.create();
      await supertest().get('/profile').expect(401);
    });
  });

  describe('PATCH /profile', () => {
    it('enables desktop notifs', async () => {
      const user = await UserFactory.create({
        desktopNotifsEnabled: false,
        phoneNotifsEnabled: false,
      });
      const res = await supertest({ userId: user.id })
        .patch('/profile')
        .send({ desktopNotifsEnabled: true })
        .expect(200);
      expect(res.body).toMatchObject({
        desktopNotifsEnabled: true,
        phoneNotifsEnabled: false,
      });
    });
    it('enables phone notifs', async () => {
      const user = await UserFactory.create({
        desktopNotifsEnabled: false,
        phoneNotifsEnabled: false,
      });
      const res = await supertest({ userId: user.id })
        .patch('/profile')
        .send({ phoneNotifsEnabled: true, phoneNumber: '911' })
        .expect(200);
      expect(res.body).toMatchObject({
        desktopNotifsEnabled: false,
        phoneNotifsEnabled: true,
      });
    });
    it('does not let student enable without phone number', async () => {
      const user = await UserFactory.create({
        desktopNotifsEnabled: false,
        phoneNotifsEnabled: false,
      });
      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ phoneNotifsEnabled: true })
        .expect(400);
    });
  });
});
