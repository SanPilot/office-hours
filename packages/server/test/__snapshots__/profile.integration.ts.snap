// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Integration GET /profile returns the logged-in user profile 1`] = `
Object {
  "courses": Array [
    Object {
      "course": Object {
        "id": 1,
        "name": "CS 2500",
      },
      "role": "student",
    },
  ],
  "desktopNotifs": Array [],
  "desktopNotifsEnabled": false,
  "email": "user@neu.edu",
  "firstName": "User",
  "id": 1,
  "insights": null,
  "lastName": null,
  "name": "User",
  "phoneNotifsEnabled": false,
  "photoURL": null,
}
`;
